dynamodb/data_test.go

```diff
@@ -0,0 +1,177 @@
+		scanOutputs: scanOutputs,
+	}
+
+	tableName := "testtable"
+	cols := []string{"a", "b", "c", "d"}
+	conv := buildConv(
+		ddl.CreateTable{
+			Name:     tableName,
+			ColNames: cols,
+			ColDefs: map[string]ddl.ColumnDef{
+				"a": {Name: "a", T: ddl.Type{Name: ddl.String, Len: ddl.MaxLength}},
+				"b": {Name: "b", T: ddl.Type{Name: ddl.Numeric}},
+				"c": {Name: "c", T: ddl.Type{Name: ddl.String, Len: ddl.MaxLength}},
+				"d": {Name: "d", T: ddl.Type{Name: ddl.Bool}},
+			},
+			Pks: []ddl.IndexKey{{Col: "a"}},
+		},
+		schema.Table{
+			Name:     tableName,
+			ColNames: cols,
+			ColDefs: map[string]schema.Column{
+				"a": {Name: "a", Type: schema.Type{Name: typeString}},
+				"b": {Name: "b", Type: schema.Type{Name: typeNumber}},
+				"c": {Name: "c", Type: schema.Type{Name: typeNumberString}},
+				"d": {Name: "d", Type: schema.Type{Name: typeBool}},
+			},
+			PrimaryKeys: []schema.Key{{Column: "a"}},
+		},
+	)
+	conv.SetDataMode()
+	var rows []spannerData
+	conv.SetDataSink(
+		func(table string, cols []string, vals []interface{}) {
+			rows = append(rows, spannerData{table: table, cols: cols, vals: vals})
+		})
+	ProcessData(conv, client, false)
+	assert.Equal(t,
+		[]spannerData{
+			{
+				table: tableName,
+				cols:  cols,
+				vals:  []interface{}{"str-1", *numVal1, "12.34", true},
+			},
+			{
+				table: tableName,
+				cols:  cols,
+				vals:  []interface{}{"str-2", *numVal2, nil, nil},
+			},
+		},
+		rows,
+	)
+}
+
+func TestCvtColValue(t *testing.T) {
+	str := "str-1"
+	numStr := "1234.56789"
+	boolVal := true
+	binaryVal := []byte("ABC")
+	binarySetVal := [][]byte{binaryVal}
+	listVal := []*dynamodb.AttributeValue{
+		{S: &str},
+		{N: &numStr},
+	}
+	mapVal := map[string]*dynamodb.AttributeValue{
+		"list": {L: listVal},
+	}
+	stringSetVal := []*string{&str}
+	numVal := big.NewRat(123456789, 100000)
+
+	testcases := []struct {
+		name    string
+		srcType string                   // Source DB type.
+		spType  string                   // Spanner DB type.
+		in      *dynamodb.AttributeValue // Input value for conversion.
+		want    interface{}              // Expected result.
+	}{
+		{"bool", typeBool, ddl.Bool, &dynamodb.AttributeValue{BOOL: &boolVal}, true},
+		{"binary", typeBinary, ddl.Bytes, &dynamodb.AttributeValue{B: binaryVal}, binaryVal},
+		{"binary set", typeBinarySet, ddl.Bytes, &dynamodb.AttributeValue{BS: binarySetVal}, binarySetVal},
+		{"map", typeMap, ddl.String, &dynamodb.AttributeValue{M: mapVal}, "{\"B\":null,\"BOOL\":null,\"BS\":null,\"L\":null,\"M\":{\"list\":{\"B\":null,\"BOOL\":null,\"BS\":null,\"L\":[{\"B\":null,\"BOOL\":null,\"BS\":null,\"L\":null,\"M\":null,\"N\":null,\"NS\":null,\"NULL\":null,\"S\":\"str-1\",\"SS\":null},{\"B\":null,\"BOOL\":null,\"BS\":null,\"L\":null,\"M\":null,\"N\":\"1234.56789\",\"NS\":null,\"NULL\":null,\"S\":null,\"SS\":null}],\"M\":null,\"N\":null,\"NS\":null,\"NULL\":null,\"S\":null,\"SS\":null}},\"N\":null,\"NS\":null,\"NULL\":null,\"S\":null,\"SS\":null}"},
```

Comment:
I see what you mean about how verbose the conversion is here.
Let's chat about this -- I'd like to understand this a bit more.
I guess one option could be to have a flag that suppresses all of the null fields to give a more useful Spanner representation.